File: explicit-java/src/main/kotlin/explicit/api/Navigator.kt
/*
 * Copyright (c) 2016-2018, Leftshift One
 * __________________
 * [2018] Leftshift One
 * All Rights Reserved.
 * NOTICE:  All information contained herein is, and remains
 * the property of Leftshift One and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Leftshift One
 * and its suppliers and may be covered by Patents,
 * patents in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Leftshift One.
 */

package explicit.api

import java.util.*
import java.util.concurrent.atomic.AtomicInteger

/**
 * Utility class for Iterator instances.
 *
 * @since 1.0.0
 */
class Navigator<T>(val list: List<T>) : Iterator<T>, Iterable<T> {

    private val index = AtomicInteger()

    override fun hasNext() = index.get() < list.size

    override fun next() = list[index.getAndIncrement()]

    fun prev() = index.getAndDecrement()

    fun reset() = index.set(0)

    fun getIndex() = index.get()

    fun setIndex(index: Int) = this.index.set(index)

    fun getCurr() = list[index.get() - 1]

    fun getPrev(): Optional<T> {
        if (index.get() >= 2)
            return Optional.of(list[index.get() - 1])
        return Optional.empty()
    }

    fun getRemaining() = list.size - index.get()

    override fun iterator(): Iterator<T> {
        return object : Iterator<T> {
            private var index = 0
            override fun hasNext() = index < list.size
            override fun next() = list[index++]
        }
    }
}

File: explicit-python/explicit_nlu/api/Navigator.py
from typing import TypeVar, Generic, List, Optional

T = TypeVar("T")


class Navigator(Generic[T]):

    def __init__(self, list: List[T]):
        self.index = 0
        self.list = list

    def __iter__(self):
        return self

    def __next__(self):
        if self.index < len(self.list):
            the_next = self.list[self.index]
            self.index += 1
            return the_next
        raise StopIteration()

    def __len__(self) -> int:
        return len(self.list)

    def has_next(self):
        return self.index < self.__len__()

    def prev(self):
        self.index = self.index - 1

    def reset(self):
        self.index = 0

    def get_index(self) -> int:
        return self.index

    def get_curr(self) -> T:
        return self.list[self.index - 1]

    def get_prev(self) -> Optional[T]:
        if self.index >= 2:
            return self.list[self.index - 2]
        return None

    def get_remaining(self) -> int:
        return self.__len__() - self.index

    def get_absolute_index(self) -> int:
        index = 0
        for i in range(self.index):
            index = index + 1
        return index
